### ISSUE TITLE
core functions: retrieve()

### ISSUE THREAD

**Issue by cboettig** [cboettig/contenturi#15]
I think retrieve may need the most input.  It's essentially a thin wrapper around `query` that returns a single location instead of a `data.frame` of all matching identifiers / URLs, but maybe I've overshot and included `pins()` style logic.  Still, it's not clear to me how to have this function 'do less' while still delivering on the promise of giving me back a single URL.  

One way `retrieve` might be cleaner is if it did not take both content hash identifier and URLs as arguments, since the logic involved is clearly different.  To me, this is tied to the same question for `query()`, see #14.  Perhaps we need separate verbs for query or retrieve  of a content hash identifier then we do for a URL?

Or is it okay for `query` to take either a URL or a content identifier, but `retrieve` should be work only with content identifiers?  (This would still require potentially downloading multiple URLs to verify though, if no local stored content is registered)

Note that  `retrieve("http://cdiac.ornl.gov/ftp/trends/co2/vostok.icecore.co2")` is still only running `query` an no other core function, that is, it is only looks for *existing registry entries*. This means that if the content returned does not match *the most recent registered content identifier at that URL*, it will not return a match.  Perhaps that is functionality that belongs only to `pin()`? (alternately you could imagine `pin()` would register new content?)

Current spec:

```r
## retrieve by hash
retrieve("hash://sha256/9412325831dab22aeebdd674b6eb53ba6b7bdd04bb99a4dbb21ddff646287e37")

## retrieve by URI
retrieve("http://cdiac.ornl.gov/ftp/trends/co2/vostok.icecore.co2")
```

It lets the user specify `prefer = c("local", "remote")` indicating that local storage entries should be tried (first, or second) to URLs. If there are multiple entries, it will go by most recent first.  `prefer` can be set in different order, or can insist on `remote` or `local` only.  (perhaps this should be more general to have preferences to different repositories, but I'm not sure, that could complicate the user interface too much?)

If remote storage is preferred, or no local store is available, `retrieve()`, will download the object and by default (`verify = TRUE`) validate that the downloaded object matches the registered hash for that URL.  If not, it will proceed to try the next entry (though currently without any message /warnings).  

Currently, for speed, `verify_local = FALSE`, meaning locally stored content is not validated (since it is stored by hash anyway.)  

(Also, it seems not ideal that this function does the downloading internally, rather than return download urls to the user, but I don't see a way around that while meeting the spec above.BUT, downloading the content seems to be a necessary step if you are given a content hash and you need to determine which of multiple URLs contain that hash.).

**Comment by cboettig** [comment]
Just adding context: @jhpoelen writes in #9: 

> It my mind, the retrieve only applies to the local store.
> 
> The flow would be:
> 
> 1. User queries one or more registries
> 2. Registries return a list of content hash generation event records (needs a bettter name!)
> 3. User chooses the most suitable location to get the content from
> 4. On retrieving the content, it can be store locally, using the store_local method.
> 
> It feels like the retrieve method is mixing-in the functionality we envisioned in the "pin" method.

Currently, this flow (well, 1 & 2, since 3 & 4 are left up to the user) are what happens with `query()`, see #14. (Though the above flow does not explicitly state if the query is only for a content hash, only for a URL, or either one).  Also, you say "only the local store", which would make this behavior slightly different from the current `query()` in that the data.frame would be filtered for only local store `source`.  I don't think it makes sense to see `retrieve` as a `query()` that is filtered for local `source` though.  

I do think there is a role for a function that can return a single source rather than a list of sources, but perhaps that is just out of scope for core functionality and we should just have `query()`?  Or did you have a different vision for the behavior of `query()`?

**Comment by cboettig** [comment]
Revised definition of `retrieve()`:

- Argument must be a content uri
- return object is a local path to content identified by the hash.

Internally, this means:
 - `query()` the content uri with the requested registries (e.g. default is all three: local `store` registry, currently formatted as a bagit manifest, local url registry, currently stored as `registry.tsv.gz`, and the hash-archive.org registry).
- query returns all locations with the requested content, so all `retrieve` has to do is choose one, download it if necessary, and validate that it matches the hash.  It does so by starting looking for content in the local `store()`, and then working through registered URLs.

**Comment by jhpoelen** [comment]
Great to see the activities on the store and registry interfaces. 

In your current definition, a ```retrieve``` finds location for content hash uri . And that sounds much like what a registry does: finding locations that are known to produce content with a specific identifier. Intuitively (in my mind), a store allows for retrieving content, and content is expressed in bits 'n bytes. This is why I'd expect a ```retrieve``` to take the registered location as an argument, to return a stream of bits and byes. This would allow any file system or web server (e.g., zenodo, figshare) to serve as act as a store.

I think the phrase: ```store registry``` highlights the mixing of the concepts. 

I imagine that you can serialize/store a local registry in various formats, including the bagit format and the registry.tsv.gz format.  I can see why the bagit format appeals because it defines a folder structure, some files names and allows for relative references pointing into a data folder. Semantically, they are pretty similar in that they provide only semantics for describing a single asset and leave out the provenance. I'll leave the provenance connection for another thread. 

So, my question is: what is the difference between a store and a registry?

**Comment by jhpoelen** [comment]
Is it correct that with the proposed definition of ```retrieve``` you mandate that every store is a content-addressed store?

**Comment by jhpoelen** [comment]
I do think that the proposed definition is more like the behavior I associated with ```pin```. This is because the operation has logic in it that makes decisions for the users on where to query and retrieve content.

**Comment by cboettig** [comment]
> So, my question is: what is the difference between a store and a registry?

A store stores content (bits and bytes), a registry only maps between identifiers and locations.   Though I have a follow-up question: Does all content live in a "store" of some form?  To my mind,   there are locations (specifically, URLs) at which we can register content but I would not consider that content to be sitting in a `store`.  Do you agree with that?  If so, should retrieve() be able to get that content, or does retrieve() only get content in a store?  

But this focuses on the nouns (store and registry).  I think we would get more clarity if we talked about the verbs explicitly, e.g.

> what a registry does: finding locations that are known to produce content with a specific identifier.

I would re-phrase this: Running `query()` on (one or more) registries finds locations known to have had content matching a specific identifier.  But that's not the only verb that touches the registry: e.g.`register()` adds identifiers+location pairs into the registry.  

I think we agree on the above, though arguably we should use different verbs to `query()` by content identifier vs `query()` by location.  (e.g. hash-archive.org calls the latter `api/history` and the former `api/sources`, but we call them both query).   Is that accurate?

> Is it correct that with the proposed definition of retrieve you mandate that every store is a content-addressed store?

no.  The proposed definition of `retrieve()` doesn't need a store at all, because it can retrieve content from URLs listed in a registry.  e.g. I imagine that `retrieve()` accesses data from Zenodo by taking a identifier as input, finding a Zenodo URL in the registry, and downloading the content.  I would not say that Zenodo is a "content-addressed store", and I would not say Zenodo is even acting as a "store" here.  If we want to say it's a store, I'm okay with that, but it means defining two different classes of 'stores', because not all stores can provide download URLs.  

In my implementation, a store is defined as anything that preserve my bits and bytes `store("myfile")`, and give them back to me when I ask for them by identifier (`retrieve(identifier)`).  The store could be redis, berkeley db, a hard disk, remote server, etc; it need not be doing any content-based-storage on the back end, but it does need to treat the identifier as a key.  

I'm not sure what it means to `retrieve()` using a URL instead of an identiifer, e.g.  `retrieve("https://example.com/data")`.  How is that different than just `download.file()`?  Sure, that URL could be registered, but what if the registry contains multiple different content at that URL?  (Recall we already have `query()` that gives all content that has been observed at that URL, but that's returning a list of identifiers, not retrieving bits and bytes.  arguably query(url) should be called `getHistory` or something)

**Comment by jhpoelen** [comment]
> A store stores content (bits and bytes), a registry only maps between identifiers and locations. Though I have a follow-up question: Does all content live in a "store" of some form? To my mind, there are locations (specifically, URLs) at which we can register content but I would not consider that content to be sitting in a store. Do you agree with that? If so, should retrieve() be able to get that content, or does retrieve() only get content in a store?

By definition, digital content lives in one or more (digital) content stores. These content stores can be accessed in various ways (e.g., using absolute/relative file paths on a hard disk, using http get on some URL of a webserver), Registries allow for keeping track of what content (identified by content hash uri) was located where in some store at some point in time. So, I'd say that anything serving content (e.g., hard disk, web server, napkin with 0's and 1' on them) is a content store. So, you can register content even if you don't own the store. This is why I consider Zenodo a content store : they provide access to content via http/get and urls.

To take an extreme example, including the napkin, registry entries would look like:

| content hash uri (what) | location (where) | date (when) | registration uuid (when "id") | provenance (who/how)
| --- | --- | --- | --- | --- |
hash://sha256/123... | 101 main street, first floor, second table to the left, yellow napkin | 2020-01-20 | event uuid X | (more on provenance later)
hash://sha256/123... | file://home/user/data/dataset.zip | 2020-02-20 |  registration event uuid Y | (more on provenance later)
hash://sha256/123... | data/dataset.zip | 2020-02-20 |  registration event uuid Y | (more on provenance later)
hash://sha256/123... | https://example.org/dataset.zip | 2020-02-15 | registration event uuid Z | (more on provenance later)
hash://sha256/123... | https://zenodo.org/record/2490831/files/taxonMap.tsv.gz | 2020-02-15 | registration event uuid A | (more on provenance later)
hash://sha256/123... | sftp://example.org/dataset.zip | 2020-01-20 | registration event uuid B | (more on provenance later)

(now attempting to use the verbs to tie this together, verbs are highlighted)

Here, someone, lets say Alice, would ```store``` content on a napkin, calculate the content uri, then that same actor (or some other actor), lets say Bob, would take that location (of the napkin) along with the content uri, and ```register``` them with some registry . This registration event is referred to by its randomly generated and unique registration event uuid. 

Now, Carol looks for any content registered on 101 main street and ```queries``` the registry by location of that napkin. The registry returns the registered content uri and one or more the locations of the napkin. Now, Carol ```selects``` a single matching location, takes public transit to 101 main street, ```locates```/```finds``` the napkin and ```reads```/ ```retrieves``` the content of the napkin. As she is reading the napkin, she verifies that the content of the napkin has the same content hash uri as the napkin location that was ```registered``` by Bob. So, Carol concludes that the napkins has not been written on since Alice and Bob last worked with it. 

With other examples, public transit can be replaced with some communication protocol (http, ftp, file) and the address can be replaced with some other locator (e.g., http url, sftp url, file url).

In summary: the sequence of verbs, actors and entities:

1. Alice stores napkin at a location
2. Bob registers napkin location and content hash uri 
3. Carol queries registry for a location
4. Carol selects/finds/locates a single, specific location associated with one or more content uri hashes
5. Carol reads (or retrieves content of) the napkin

With variation 3b. where Carol queries registry for a content uri, and 4b. where Carol picks one location that was most recently registered to provide the content uri that Carol is looking for.
 
Looking back, I interpreted ```retrieve``` to be ```read``` and lumped together the ```query``` and  ```find```/```locate```/```select```` location.  This is why I advocated for ```bits 'n bytes <- retrieve(location)```. In my mind, content stores are content uri agnostic, at least as far as their "public" interfaces are concerned. 

In this scheme, ```retrieve(location)``` is similar to saying ```httpget(location)``` if the location is a http url, or similar to saying ```read(location)``` for a file url. In case of the napkin, you need a human to look at the napkin read the content: ```human_reads(napkin at some location)```. 

So, I'd say that you can ```identify``` content by their content hash uri , but need to ```retrieve``` the content from some location to have a look at it. And, by definition, this location resides on some content store (e.g., 101 main street, zenodo webservers, ftp site, etc.).

My questions: 1. does my reasoning make sense to you? 2. if so, what verbs can we used to avoid confusion (I suspect, this conversation will happen again if we leave it as is) 3. if not, please let me know what parts are inconsistent. 

Thanks for being patient with me.

**Comment by cboettig** [comment]
Thanks, this is great, and I also appreciate your patience.  

I'm happy to adopt the terminology that all content *locations* are content *stores*, though this has some important implementation implications we will need to work through.  

I like the narrative here, it closely matches `retrieve()` and `query()` as implemented, but there's one part I'm still stuck on:

> Now, some time later, Carol looks for any content registered on 101 main street and queries the registry by location of that napkin. The registry returns the registered content uri and one or more the locations of the napkin.

This seems like it could blow up?
- "the registered content uri"  -- but there may be many many different content URIs that have been registered at that location.  The description above implies that this action is really at least 2 actions in, say, hash-archive.org verbs.  For a single location, I could get lots (say, `n`) of content URIs, then I would have to query all of those content URIs, each of which may also be at `m_i` different locations.    The narrative is confusing when Carol queries by location because (a) she doesn't learn anything from this action.  She queries by 101 main street, she gets back the location 101 main street (and possibly other locations?)  She decides then to go to 101 main street.  What was the point of the query?  

the story makes more sense to me with 3b, 4b, variation.  


This narrative is I think pretty consistent with the current functions of `register()` and `query()`, and I think we're on the same page there.  

In this scheme, `retrieve()` sounds like a nice generic term to refer to a collection of protocol-specific verbs (download, read, or other protocols, like get from SQL, Redis, or Berkeley DB key-value store).  But it also does not sound like something that would be an actual protocol we would define as a function inside this package, but rather a scope that is already covered by the existing protocol-specific verbs of `readr` and so many other packages.  Yes?

Meanwhile, the core functionality of the package just stops at `query()` returning a data.frame of all hashes and locations (modulo my question above re 3a 4a path).  

Also, we have the function (`store()`, but maybe that's a poor verb too, because really it refers to a very specific kind of content-addressed local `store()` only.  

To my mind, there is an obvious protocol for the task of returning bits & bytes (or rather, a local path to those bits and bytes) given a content URI, which involves the process I outlined above and called `retrieve`.  We cannot call that process `pin()` because it does not behave like `pins::pin()`, which takes a URL, stores a persistent cache, and returns the bits & bytes.  My operation does not take a URL, and does not create any persistent store, so it is not `pin()`.  You could easily make it `pin()`, you'd have to do:

```
pin <- function(url){
  id <- store(url)
  retrieve(id)
}
```

**Comment by jhpoelen** [comment]
> This seems like it could blow up?
Yes, many matching content-uri / location pairs would be returned on vague location queries or highly dynamic locations (e.g., some newsfeed or dynamic dataset). However, I don't see why you have to query for all the resulting content-uri. You can, but you don't have to. And, I think Carol does learn something new by querying for 101 main street, because she learned about the registered napkin content by their content hash uri. 

> In this scheme, retrieve() sounds like a nice generic term to refer to a collection of protocol-specific verbs (download, read, or other protocols, like get from SQL, Redis, or Berkeley DB key-value store). But it also does not sound like something that would be an actual protocol we would define as a function inside this package, but rather a scope that is already covered by the existing protocol-specific verbs of readr and so many other packages. Yes?

Ok - I like the idea that the package's scope would be to help register, and search for content, using certain location - content hash uri pairs. And, for convenience, perhaps include a local content-uri-local-store around, or introduce that in a new package content-uri-store. Creating a separate package would open up the creation of content-uri-zenodo-store , content-uri-s3-store, content-uri-dropbox-store etc. 

I agree that ```store``` does not behave like ```pin```. For some reason, I take issue with ```store(location)``` as opposed to ```store(bits 'n bytes)``` . I'd want to emphasize that we are storing  data. For the ```store```, the writing of bits is followed by a registration of a content-uri - location pair. You _store_ bytes, and _register_  store locations and their associated content hash uris. If you store by location, there's no guarantee that that location that is being stored has the content that the user intends to store. For instance, if the location points to a file, the file may have gotten altered by some other process already.

Expressing this nuance in the ```pin``` example:

```R
pin <- function(url) {
  (id, file_url) <- contenturi-filestore::store(httr::get(url), registries=my_registry_list)
  readr(file_url)
}
``` 

In the GloBI implementation, I use a [ContentSource](https://github.com/globalbioticinteractions/globalbioticinteractions/blob/212647859c7e037d1474be6f55239ec24452c8d7/eol-globi-cache/src/main/java/org/globalbioticinteractions/content/ContentSource.java#L8-L7) as a way to reference bits 'n bytes without having to directly couple it to some locator. 

So, assuming that we are on the same page (please confirm!), the question is: now that contenturi package will focus on registration verbs (```register``` / ```query```) do we keep the local content store in the contenturi package, or do we created a separate package? 

As far as ```store``` verb - it really means ```provide_and_register```,  ```save_and_register``` or ```hash_and_register```.

**Comment by cboettig** [comment]
Store already exists as a separate package, see https://cran.r-project.org/web/packages/storr/vignettes/storr.html

I believe it works in the way you describe (i.e. the objects you store are bits and bytes, not serialized files).  You can create different store backends (DBI-based, Redis based, rds based, lightning-based), 


```r
st <- storr::storr_rds(path) # RDS-backed store
st$set("mykey", mtcars)     # store the mtcars data.frame under a "user friendly" key
st$get(mykey)                     # retrieve the bits & bytes
```

Objects can also be retrieved directly from their content hash instead of their key:

```r
hash_id <- st$hash_object(mtcars)
st$get_value(hash_id)
```

By default these are md5sums, but can be sha256 sums instead.  Of course they won't be the same sha256sums you would have on the serialized file, so these content identifiers do not match content identifiers you would find by, say, passing a URL to `hash-archive.org`.  As you point out, this is different than the `contenturi`'s `store()` function that takes a location, and is thus hashing serialized file objects, not objects that have been parsed into R's memory.  I think there are practical advantages of hashing the serialized file, but also limitations so it might be something we want to think through in more detail.  

(So `storr` is using a content-based storage under the hood, though the back-end stores aren't necessarily content-addressed systems themselves).  


```

**Comment by cboettig** [comment]
Closed by #23, store and retrieve are now simple inverses operations for content id <--> bits & bytes.
